workflow/snakemake_rules/main_workflow.smk

```diff
@@ -597,6 +568,51 @@ rule compress_build_align:
         xz -c {input} > {output} 2> {log}
         """
 
+rule filter:
+    message:
+        """
+        Filtering alignment {input.sequences} -> {output.sequences}
+          - excluding strains in {input.exclude}
+          - including strains in {input.include}
+          - min length: {params.min_length}
+        """
+    input:
+        sequences = "results/{build_name}/masked.fasta",
+        metadata = "results/{build_name}/{build_name}_subsampled_metadata.tsv.xz",
+        # TODO - currently the include / exclude files are not input (origin) specific, but this is possible if we want
+        include = config["files"]["include"],
+        exclude = _collect_exclusion_files,
+    output:
+        sequences = "results/{build_name}/filtered.fasta",
+    log:
+        "logs/filtered_{build_name}.txt"
+    benchmark:
+        "benchmarks/filter_{build_name}.txt"
+    params:
+        min_length = lambda wildcards: _get_filter_value(wildcards, "min_length"),
+        exclude_where = lambda wildcards: _get_filter_value(wildcards, "exclude_where"),
+        min_date = lambda wildcards: _get_filter_value(wildcards, "min_date"),
+        ambiguous = lambda wildcards: f"--exclude-ambiguous-dates-by {_get_filter_value(wildcards, 'exclude_ambiguous_dates_by')}" if _get_filter_value(wildcards, "exclude_ambiguous_dates_by") else "",
+        date = (date.today() + datetime.timedelta(days=1)).strftime("%Y-%m-%d"),
+    resources:
+        # Memory use scales primarily with the size of the metadata file.
+        mem_mb=500
+    conda: config["conda_environment"]
+    shell:
+        """
+        augur filter \
+            --sequences {input.sequences} \
+            --metadata {input.metadata} \
```

Comment:
We could support input-specific filters as long as the metadata include columns like `length` or `ACTG` or `valid_bases` and then users could define a query that filters based on metadata input column like:

```
--query "(gisaid == 'yes' & length > 5000)"
```

If we update the alignment step above to use Nextclade instead of Nextalign, we could get the annotation of the number of gaps, etc. in a TSV to merge with metadata to allow the query above.

@jameshadfield points out that we could generate this query based on the existing filter config if users provide an input-specific filter and add that query to the existing augur filter. To do this we would loop through all "origins" defined in `inputs` and construct a query string with a component per origin that pulls from the config. For a config like the following:

```yaml
filter:
  min_length: 10000
  aus:
    min_length: 5000
```

the resulting query would be something like:

```
--query "(gisaid == 'yes' & length > 10000) | (aus == 'yes' & length > 5000)"
```

Reply:
Thinking about this more now, this solution only addresses the specific use case of filtering by length. We've technically allowed users [to define any valid filter key for any origin](https://github.com/nextstrain/ncov/blob/27cd1c251f7c066a8f18bf7c1541bfe90f698162/workflow/snakemake_rules/common.smk#L31-L35) and we've documented [four specific options in the configuration guide](https://docs.nextstrain.org/projects/ncov/en/latest/reference/configuration.html#filter) and [two of these in the multiple inputs example](https://github.com/nextstrain/ncov/blob/27cd1c251f7c066a8f18bf7c1541bfe90f698162/my_profiles/example_multiple_inputs/builds.yaml#L23-L25).

Instead of figuring out how to support each of these individually or collectively, we could revisit the use case. Here's an attempt at the user story we're trying to support:

> James has 20 new likely Omicron genomes from his city that just came off the sequencer and he wants to know how they place in the phylogenetic context of other sequences from the city and the country. He configures a Nextstrain build with inputs including the full GISAID database and his local sequences and metadata. He wants a high-quality set of GISAID data for his phylogenetic context, but he wants to keep all local sequences in his analysis regardless of their quality. He sets stringent quality filters for GISAID data, applying a minimum genome length of 27,000 valid nucleotides and filtering out any genomes that do not pass all of Nextclade's quality control annotations.

From this user's perspective, we've got the current implementation a bit backwards; James wants to keep all of his inputs by default (with reasonable subsampling) and only wants to apply filters to specific datasets. We currently require the opposite approach where we apply more stringent filters by default and require users to manually relax those filters.

Reply:
I do think having a more stringent default is likely good - for users less clear on what's going on, I think reducing the odds they take away big messages from poor quality genomes is fair. However, it should also be as straightforward as possible to relax these if you actively are choosing to include worse-quality genomes (which is a very fair thing to do - just really important you know you've done it!).

I think if most users are interested in QC metric filtering, we could try to get this from Nextclade early on (as you suggest) & support quite a few different filtering options - but not "anything". I guess we also ideally would want for user-datasets to go through this, but not to try and generate this for GISAID - preferring to wait until after subsampling (though those within Nextstrain can use the files we've generated already with this info). 

I like the idea of putting this query together from the config as suggested above.